signwriting_animation/data/data_loader.py

```diff
@@ -0,0 +1,123 @@
+import os
+import random
+import torch
+import pandas as pd
+from torch.utils.data import Dataset, DataLoader
+from pose_format.torch.masked.collator import zero_pad_collator
+from pose_format.pose import Pose
+from pose_anonymization.data.normalization import normalize_mean_std
+from signwriting.visualizer.visualize import signwriting_to_image
+from transformers import CLIPProcessor
+from PIL import Image
+
+class DynamicPosePredictionDataset(Dataset):
+    """
+    Dataset for dynamic on-the-fly sampling of normalized pose data
+    with SignWriting condition images and scalar metadata padded as length-1 tensors.
+    """
+    def __init__(self,data_dir,csv_path,past_frames=40,future_frames=20,with_metadata=True):
+        super().__init__()
+        self.data_dir = data_dir
+        self.num_past_frames = num_past_frames
+        self.num_future_frames = num_future_frames
+        self.with_metadata = with_metadata
+
+        df = pd.read_csv(csv_path)
+        self.records = df.to_dict(orient='records')
+
+        # Initialize CLIP processor for signwriting images
+        self.clip_processor = CLIPProcessor.from_pretrained("openai/clip-vit-base-patch32")
+
+    def __len__(self):
+        return len(self.records)
+
+    def __getitem__(self, idx):
+        rec = self.records[idx]
+        start = rec.get("start", 0)
+        total = rec.get(
+            "end", start + self.num_past_frames + self.num_future_frames
+        )
+        end = min(start + self.num_past_frames + self.num_future_frames, total)
+
+        # Load & normalize pose bytes
+        pose_path = os.path.join(self.data_dir, rec["pose"])
+        with open(pose_path, "rb") as f:
+            raw = f.read()
+        pose = Pose.read(raw)
+        pose = normalize_mean_std(pose)
+
+        data = pose.body.data.data           
+        mask = torch.tensor(pose.body.data.mask, dtype=torch.bool)
+
+        # Slice sequences
+        input_pose = data[max(start, 0): start + self.num_past_frames]
+        target_pose = data[start + self.num_past_frames : start + self.num_past_frames + self.num_future_frames]
+
+        # Invert masks (True=visible)
+        input_mask = torch.logical_not(mask[max(start, 0): start + self.num_past_frames])
+        target_mask = torch.logical_not(mask[start + self.num_past_frames : start + self.num_past_frames + self.num_future_frames])
+
+        # Render SignWriting to PIL
+        sw_text = rec.get("text", "")
+        if sw_text:
+            pil_img = signwriting_to_image(sw_text)
+        else:
+            pil_img = Image.new("RGB", (224, 224), color=(0, 0, 0))
```

Comment:
why would there not be signwriting text?
In any case, if that does happen, you can instead:
```py
sw_text = rec.get("text", "M500x500")
pil_img = signwriting_to_image(sw_text)
```
and avoid duplicated logic